### PR TITLE
Gkanz/import openapi

### PR DESCRIPTION
**Pull request by gkanz**
Fixes # N/A

Changes proposed in this pull request:
- Add OpenAPI 3.0 importing
- Assuming OpenAPI 2.0 importing will be maintained, requires factoring out common code.

@anz-bank/sysl-developers

### CODE REVIEW

File: src/sysl/importers/import_openapi.py [hunk 1]
```diff
@@ -0,0 +1,520 @@
...
+def main():
+    args = parse_args(sys.argv)
+
+    with open(args.openapi_path, 'r') as f:
+        oaSpec = yaml.load(f)
+
+    w = writer.Writer('sysl')
+
+    translator = OpenApiTranslator(logger=make_default_logger())
+    translator.translate(oaSpec, args.appname, args.package, w=w)
```
Comment (anzdaddy) [inline src/sysl/importers/import_openapi.py]: s/w=w/w/

File: src/sysl/importers/import_openapi.py [hunk 2]
```diff
@@ -0,0 +1,520 @@
...
+            if 'host' in oaSpec:
+                w('@host = {}', json.dumps(oaSpec['host']))
+
+            w(u'@description =:')
+            with w.indent():
+                w(u'| {}', oaSpec['info'].get('description', 'No description.').replace("\n", "\n|"))
+
+            if 'basePath' in oaSpec:
+                w()
+                w('{}:'.format(oaSpec['basePath']))
```
Comment (anzdaddy) [inline src/sysl/importers/import_openapi.py]:
```python
w('{}:', oaSpec['basePath'])
```
Same everywhere `sysl.util.Writer` is used.

File: src/sysl/importers/import_openapi.py [hunk 3]
```diff
@@ -0,0 +1,520 @@
...
+            if len(returnValues) > 0:
+                w(u'return {}'.format(', '.join(returnValues)))
+            else:
+                w(u'return')
+
+        for (path, api) in sorted(oaSpec['paths'].iteritems()):
+            apiParams = []
+            if 'parameters' in api:
+                apiParams = api['parameters']
+                del api['parameters']
```
Comment (anzdaddy) [inline src/sysl/importers/import_openapi.py]:
```python
apiParams = api.pop('parameters') if 'parameters' in api else []
```

File: src/sysl/importers/import_openapi.py [hunk 4]
```diff
@@ -0,0 +1,520 @@
...
+
+            for (i, (method, body)) in enumerate(sorted(api.iteritems(), key=lambda t: METHOD_ORDER[t[0]])):
+                bodyContent = OrderedDict()
+                bodyParams = body['parameters'] if 'parameters' in body else []
+                pathParams, headerParams, queryParams = self.mergeParams(oaSpec, apiParams, bodyParams)
+                print(pathParams)
```
Comment (anzdaddy) [inline src/sysl/importers/import_openapi.py]: Is this debug logging?

File: src/sysl/importers/import_openapi.py [hunk 5]
```diff
@@ -0,0 +1,520 @@
...
+                w(u'\n{}:', self.translate_path_template_params(path, pathParams))
+                with w.indent():
+                    header = self.getHeaders(headerParams)
+                    jsonBody = []
+                    if body.get('requestBody', None):
+                        if body['requestBody'].get('content', None):
+                            if body['requestBody']['content'].get('application/json', None):
+                                jsonBody.append(body['requestBody']['content']['application/json'])
```
Comment (anzdaddy) [inline src/sysl/importers/import_openapi.py]:
```python
appJson = body.get('requestBody', {}).get('content', {}).get('application/json', None)
jsonBody = [appJson] if appJson else []
```

File: src/sysl/importers/import_openapi.py [hunk 6]
```diff
@@ -0,0 +1,520 @@
...
+                word = m.group(1)
+                if word in self.words() and word + 'id' not in self.words():
+                    ident = word + 'Id'
+            self._param_cache[param] = ident
+
+        return ident
+
+    def translate_path_template_params(self, path, params):
+        parts = re.split(r'({[^/]*?})', path)
+        if len(parts) > 1 and len(params) == 0:
```
Comment (anzdaddy) [inline src/sysl/importers/import_openapi.py]:
```python
if len(parts) > 1 and not params:
```

File: src/sysl/importers/import_openapi.py [hunk 7]
```diff
@@ -0,0 +1,520 @@
...
+                        if body['requestBody'].get('content', None):
+                            if body['requestBody']['content'].get('application/json', None):
+                                jsonBody.append(body['requestBody']['content']['application/json'])
+                    methodBody = self.getBody(jsonBody)
+                    if len(header) > 0 and len(methodBody) > 0:
+                        paramStr = ' ({})'.format(methodBody + ', ' + header)
+                    elif len(header) == 0 and len(methodBody) == 0:
+                        paramStr = ''
+                    else:
+                        paramStr = ' ({})'.format(methodBody + header)
```
Comment (anzdaddy) [inline src/sysl/importers/import_openapi.py]:
```python
params = [p for p in [methodBody, header] if p]
paramStr = ' ({})'.format(', '.join(params)) if params else ''
```

File: src/sysl/importers/import_openapi.py [hunk 8]
```diff
@@ -0,0 +1,520 @@
...
+
+    def writeDefs(self, oaSpec, w):
+        w()
+        w('#' + '-' * 75)
+        w('# definitions')
+
+        if not oaSpec.get('components', None):
+            return
+        if not oaSpec['components'].get('schemas', None):
+            return
```
Comment (anzdaddy) [inline src/sysl/importers/import_openapi.py]:
```python
schemas = oaSpec.get('components', {}).get('schemas', None)
if not schemas:
    return
```

File: src/sysl/importers/import_openapi.py [hunk 9]
```diff
@@ -0,0 +1,520 @@
...
+
+        def _collectExternalsSoReferencesAreCorrect():
+            for (tname, tspec) in sorted(oaSpec['components']['schemas'].iteritems()):
+                properties = extract_properties(tspec)
+
+                if properties or (tspec.get('type') == 'array') or ('enum' in tspec):
```
Comment (anzdaddy) [inline src/sysl/importers/import_openapi.py]: Redundant parens.

File: src/sysl/importers/import_openapi.py [hunk 10]
```diff
@@ -0,0 +1,520 @@
...
+                    continue
+                externAlias['EXTERNAL_' + tname] = 'string'
+
+        _collectExternalsSoReferencesAreCorrect()
+
+        for (tname, tspec) in sorted(oaSpec['components']['schemas'].iteritems()):
```
Comment (anzdaddy) [inline src/sysl/importers/import_openapi.py]:
```python
sorted(schemas.iteritems())
```

File: src/sysl/importers/import_openapi.py [hunk 11]
```diff
@@ -0,0 +1,520 @@
...
+                        fname if fname not in SYSL_TYPES else fname + '_',
+                        ftypeSyntax)
+
+                if properties:
+                    for (fname, fspec) in sorted(properties.iteritems()):
+                        (ftype, fdescr) = self.parse_typespec(fspec, fname, tname)
+                        w(getfields(fname, ftype))
+                        with w.indent():
+                            w(self.getTag(fname, 'json_tag'))
+                            if fdescr is not None and len(fdescr) > 0:
```
Comment (anzdaddy) [inline src/sysl/importers/import_openapi.py]:
```python
if fdescr:
```

File: src/sysl/importers/import_openapi.py [hunk 12]
```diff
@@ -0,0 +1,520 @@
...
+                                w(self.getTag(fdescr, 'description'))
+                # handle top-level arrays
+                elif tspec.get('type') == 'array':
+                    (ftype, fdescr) = self.parse_typespec(tspec, '', tname)
+                    w(getfields('', ftype, True))
+                    with w.indent():
+                        if fdescr is not None and len(fdescr) > 0:
```
Comment (anzdaddy) [inline src/sysl/importers/import_openapi.py]:
```python
if fdescr:
```

File: src/sysl/importers/import_openapi.py [hunk 13]
```diff
@@ -0,0 +1,520 @@
...
+                            w(self.getTag(fdescr, 'description'))
+                elif 'enum' in tspec:
+                    w(TYPE_MAP[tspec.get('type')])
+
+        index = 0
+        while len(typeSpecList) > index:
```
Comment (anzdaddy) [inline src/sysl/importers/import_openapi.py]:
```python
for typeSpec in typeSpecList:
```

File: src/sysl/importers/import_openapi.py [hunk 14]
```diff
@@ -0,0 +1,520 @@
...
+    def translate(self, oaSpec, appname, package, w):
+        hasInfo = False
+
+        hasInfo = 'info' in oaSpec
+        title = oaSpec['info'].pop('title', '') if hasInfo else ''
+
+        w(u'{}{} [package={}]:',
+            appname, title and ' ' + json.dumps(title), json.dumps(package))
+
+        with w.indent():
```
Comment (anzdaddy) [inline src/sysl/importers/import_openapi.py]:
```python
with w.indent(u'{}{} [package={}]:', appname, title and ' ' + json.dumps(title), json.dumps(package)):
```
Same everywhere else `Writer.indent()` is used.

File: src/sysl/importers/import_openapi.py [hunk 15]
```diff
@@ -0,0 +1,520 @@
...
+                w('!type {}:', typeName)
+                with w.indent():
+                    for (k, v) in extract_properties(typeSpecList[index].element).iteritems():
+                        typeRef = typeSpecList[index].typeRef + '_' + typeSpecList[index].parentRef
+                        ftypeSyntax = self.parse_typespec(v, k, typeRef)[0]
+                        if 'required' not in typeSpecList[index].element or k not in typeSpecList[index].element['required']:
+                            ftypeSyntax = ftypeSyntax + '?'
+                        fields = '{} <: {}:'.format(
+                            k if k not in SYSL_TYPES else k + '_',
+                            ftypeSyntax)
```
Comment (anzdaddy) [inline src/sysl/importers/import_openapi.py]: Single line

File: src/sysl/importers/import_openapi.py [hunk 16]
```diff
@@ -0,0 +1,520 @@
...
+
+        # skip invalid arrays
+        if not typ and 'items' in tspec:
+            self.warn('Ignoring unexpected "items". Schema has "items" but did not have defined "type". Note: %r' % (tspec, ))
+            del tspec['items']
+
+        descr = tspec.pop('description', None)
+
+        if typ == 'array':
+            assert not (set(tspec.keys()) - {'type', 'items', 'example'}), tspec
```
Comment (anzdaddy) [inline src/sysl/importers/import_openapi.py]: Redundant outer parens

File: src/sysl/importers/import_openapi.py [hunk 17]
```diff
@@ -0,0 +1,520 @@
...
+
+        if ref:
+            assert not set(tspec.keys()) - {'$ref'}, tspec
+            t = ref.replace('#/components/schemas/', '')
+            if externAlias.get('EXTERNAL_' + t, None):
+                return r('EXTERNAL_' + t)
+            if externAlias.get('EXTERNAL_' + t + '_obj', None):
+                return r('EXTERNAL_' + t + '_obj')
+            return r(t)
+        if (typ, fmt) == ('string', None):
```
Comment (anzdaddy) [inline src/sysl/importers/import_openapi.py]: Use a mapping dict.

File: test/importers/test_import_openapi.py [hunk 18]
```diff
@@ -0,0 +1,1289 @@
...
+    swag = yaml.load(input)
+    w = writer.Writer('sysl')
+    logger = FakeLogger()
+    t = OpenApiTranslator(logger)
+    t.translate(swag, appname='', package='', w=w)
+    return str(w), logger
+
+
+def test_importing_simple_openapi_with_json_tags():
+    output, _ = getOutputString(SIMPLE_OPENAPI_EXAMPLE)
```
Comment (anzdaddy) [inline test/importers/test_import_openapi.py]:
```python
def test_importing_simple_openapi_with_json_tags():
    output, _ = getOutputString(r"""\
        "openapi": "3.0"
        info:
          title: Simple
        paths:
          /test:
            get:
        ⋮
    """)
```

File: test/importers/test_import_openapi.py [hunk 19]
```diff
@@ -0,0 +1,1289 @@
...
+                | No description.
+                return
+
+    #---------------------------------------------------------------------------
+    # definitions
+"""
+
+
+def getOutputString(input):
+    swag = yaml.load(input)
```
Comment (anzdaddy) [inline test/importers/test_import_openapi.py]:
```python
prefix = re.match('^ *', input).group(0)
swag = yaml.load(re.sub(r'^' + prefix, '', input))
```

File: test/importers/test_import_openapi.py [hunk 20]
```diff
@@ -0,0 +1,1289 @@
...
+
+class FakeLogger:
+    def __init__(self):
+        self.warnings = []
+
+    def warn(self, msg):
+        self.warnings.append(msg)
+
+
+SIMPLE_OPENAPI_EXAMPLE = r"""
```
Comment (anzdaddy) [inline test/importers/test_import_openapi.py]: Inline single-use strings.

File: test/importers/test_import_openapi.py [hunk 21]
```diff
@@ -0,0 +1,1289 @@
...
+    w = writer.Writer('sysl')
+    logger = FakeLogger()
+    t = OpenApiTranslator(logger)
+    t.translate(swag, appname='', package='', w=w)
+    return str(w), logger
+
+
+def test_importing_simple_openapi_with_json_tags():
+    output, _ = getOutputString(SIMPLE_OPENAPI_EXAMPLE)
+    assert 'name <: string?:\n            @json_tag = "name"' in output
```
Comment (anzdaddy) [inline test/importers/test_import_openapi.py]: Should also validate correct relative indent.

File: src/sysl/importers/import_openapi.py [hunk 22]
```diff
@@ -0,0 +1,520 @@
...
+    return 'sequence of ' + itemtype
+
+
+def is_sysl_array_type(ftype):
+    return ftype.startswith('sequence of ')
+
+
+def type_as_key(swagt):
+    if isinstance(swagt, dict):
+        return frozenset(sorted(swagt.iteritems()))
```
[the remark this answers is not in the document]
Comment (gkanz) [reply src/sysl/importers/import_openapi.py]: No, and that function is never called. It has been removed.

File: src/sysl/importers/import_openapi.py [hunk 23]
```diff
@@ -0,0 +1,520 @@
...
+
+def load_vocabulary(words_fn='/usr/share/dict/words'):
+    if not os.path.exists(words_fn):
+        return []
+    else:
+        return (w.strip() for w in open(words_fn))
+
+
+class OpenApiTranslator:
+    def __init__(self, logger, vocabulary_factory=None):
```
[the remark this answers is not in the document]
Comment (gkanz) [reply src/sysl/importers/import_openapi.py]: Fixed.

File: src/sysl/importers/import_openapi.py [hunk 24]
```diff
@@ -0,0 +1,520 @@
...
+
+    def words(self):
+        """Lazy-load WORDS."""
+        if not self._words:
+            self._words.update(self._vocabulary_factory())
+            if not self._words:
+                self.warn("could not load any vocabulary, janky environment-specific heuristics for renaming path template names may fail")
+        return self._words
+
+    def javaParam(self, param):
```
[the remark this answers is not in the document]
Comment (gkanz) [reply src/sysl/importers/import_openapi.py]: rfc thinks the approach for making renaming path template names is janky...

File: src/sysl/importers/import_openapi.py [hunk 25]
```diff
@@ -0,0 +1,520 @@
...
+        # TODO(anz-rfc) this seems janky and fragile.
```
[the remark this answers is not in the document]
Comment (gkanz) [reply src/sysl/importers/import_openapi.py]: rfc thinks matching against known english words to detect trailing ID is janky... no change.

File: src/sysl/importers/import_openapi.py [hunk 26]
```diff
@@ -0,0 +1,520 @@
...
+            if parts[i] in pathParams:
+                param = pathParams[parts[i]]
+                parts[i] = "{" + self.javaParam(parts[i]) + "<:" + TYPE_MAP[param['schema']['type']] + "}"
+            elif parts[i] != "" and parts[i][0] == "{":
+                self.warn("could not find type for path param: %s in params%s" % (parts[i], params))
+                parts[i] = "{" + self.javaParam(parts[i]) + "<:string}"
+        return "".join(parts)
+
+    def translate(self, oaSpec, appname, package, w):
+        hasInfo = False
```
[the remark this answers is not in the document]
Comment (gkanz) [reply src/sysl/importers/import_openapi.py]: Removed.

File: src/sysl/importers/import_openapi.py [hunk 27]
```diff
@@ -0,0 +1,520 @@
...
+                with w.indent():
+                    self.writeEndpoints(oaSpec, w)
+            else:
+                self.writeEndpoints(oaSpec, w)
+            self.writeDefs(oaSpec, w)
+
+    def mergeParams(self, oaSpec, srcParams, overrideParams):
+        pathParams = OrderedDict()
+        headerParams = OrderedDict()
+        queryParams = OrderedDict()
```
[the remark this answers is not in the document]
Comment (gkanz) [reply src/sysl/importers/import_openapi.py]: Changed.

File: src/sysl/importers/import_openapi.py [hunk 28]
```diff
@@ -0,0 +1,520 @@
...
+                param = oaSpec['parameters'][param['$ref'].rpartition('/')[2]]
+            paramIn = param['in']
+            if paramIn == 'path':
+                pathParams[param["name"]] = param
+            elif paramIn == 'header':
+                headerParams[param["name"]] = param
+            elif paramIn == 'query':
+                queryParams[param["name"]] = param
+            else:
+                raise "Unknown param type"
```
[the remark this answers is not in the document]
Comment (gkanz) [reply src/sysl/importers/import_openapi.py]: Changed.

File: src/sysl/importers/import_openapi.py [hunk 29]
```diff
@@ -0,0 +1,520 @@
...
+
+        return pathParams.values(), headerParams.values(), queryParams.values()
```
[the remark this answers is not in the document]
Comment (gkanz) [reply src/sysl/importers/import_openapi.py]: Changed.

File: src/sysl/importers/import_openapi.py [hunk 30]
```diff
@@ -0,0 +1,520 @@
...
+                    else:
+                        returnValues[schema['$ref'].rpartition('/')[2]] = True
+
+                if rc == 'default' or rc.startswith('x-'):
+                    self.warn('default responses and x-* responses are not implemented')
+
+            if len(returnValues) > 0:
+                w(u'return {}'.format(', '.join(returnValues)))
+            else:
+                w(u'return')
```
[the remark this answers is not in the document]
Comment (gkanz) [reply src/sysl/importers/import_openapi.py]: It didn't... but it does now.

File: src/sysl/importers/import_openapi.py (hunk 1 above)
Comment (gkanz) [reply src/sysl/importers/import_openapi.py]: Changed.

File: src/sysl/importers/import_openapi.py (hunk 2 above)
Comment (gkanz) [reply src/sysl/importers/import_openapi.py]: Changed, that was the only instance of `w\(.*\.format` I found.

File: src/sysl/importers/import_openapi.py (hunk 3 above)
Comment (gkanz) [reply src/sysl/importers/import_openapi.py]: Changed.

File: src/sysl/importers/import_openapi.py (hunk 4 above)
Comment (gkanz) [reply src/sysl/importers/import_openapi.py]: Removed.

File: src/sysl/importers/import_openapi.py (hunk 6 above)
Comment (gkanz) [reply src/sysl/importers/import_openapi.py]: Changed.

File: src/sysl/importers/import_openapi.py (hunk 7 above)
Comment (gkanz) [reply src/sysl/importers/import_openapi.py]: Changed.

File: src/sysl/importers/import_openapi.py (hunk 8 above)
Comment (gkanz) [reply src/sysl/importers/import_openapi.py]: Changed and substituted occurrences of `oaSpec['components']['schemas']` for `schemas`.

File: src/sysl/importers/import_openapi.py (hunk 9 above)
Comment (gkanz) [reply src/sysl/importers/import_openapi.py]: Removed.

File: src/sysl/importers/import_openapi.py (hunk 10 above)
Comment (gkanz) [reply src/sysl/importers/import_openapi.py]: Changed.

File: src/sysl/importers/import_openapi.py (hunk 11 above)
Comment (gkanz) [reply src/sysl/importers/import_openapi.py]: Changed.

File: src/sysl/importers/import_openapi.py (hunk 12 above)
Comment (gkanz) [reply src/sysl/importers/import_openapi.py]: Changed.

File: src/sysl/importers/import_openapi.py (hunk 13 above)
Comment (gkanz) [reply src/sysl/importers/import_openapi.py]: Changed and made appropriate substitutions.

File: src/sysl/importers/import_openapi.py (hunk 14 above)
Comment (gkanz) [reply src/sysl/importers/import_openapi.py]: Changed everywhere.

File: src/sysl/importers/import_openapi.py (hunk 15 above)
Comment (gkanz) [reply src/sysl/importers/import_openapi.py]: Changed.

File: src/sysl/importers/import_openapi.py (hunk 16 above)
Comment (gkanz) [reply src/sysl/importers/import_openapi.py]: Changed.

File: src/sysl/importers/import_openapi.py (hunk 17 above)
Comment (gkanz) [reply src/sysl/importers/import_openapi.py]: Changed.

File: test/importers/test_import_openapi.py (hunk 18 above)
Comment (gkanz) [reply test/importers/test_import_openapi.py]: Changed and deleted `SIMPLE_OPENAPI_EXAMPLE`. Why don't you want all the examples in one place?

File: test/importers/test_import_openapi.py (hunk 19 above)
Comment (gkanz) [reply test/importers/test_import_openapi.py]: Changed to:
```python
prefix = re.match('^ *', input).group(0)
swag = yaml.load(re.sub(r'^' + prefix, '', input), Loader=yaml.FullLoader)
```
Why not `lstrip()`? or `re.sub(r'^ *', '', input)`?

File: test/importers/test_import_openapi.py (hunk 20 above)
Comment (gkanz) [reply test/importers/test_import_openapi.py]: Really? OK, Done.

File: test/importers/test_import_openapi.py (hunk 21 above)
Comment (gkanz) [reply test/importers/test_import_openapi.py]: Done.

**Pull request merged**